lib/internal/backend/repository/repository2.go

```diff
@@ -167,48 +175,156 @@ func (rb *RepositoryBackend) buildACIV2(layerIDs []string, dockerURL *types.Pars
 	return aciLayerPaths, aciManifests, nil
 }
 
-func (rb *RepositoryBackend) getManifestV2(dockerURL *types.ParsedDockerURL) (*v2Manifest, []string, error) {
+type layer struct {
+	index  int
+	file   *os.File
+	closer io.Closer
+	err    error
+}
+
+func (rb *RepositoryBackend) buildACIV22(layerIDs []string, dockerURL *types.ParsedDockerURL, outputDir string, tmpBaseDir string, compression common.Compression) ([]string, []*schema.ImageManifest, error) {
+	layerFiles := make([]*os.File, len(layerIDs))
+
+	tmpParentDir, err := ioutil.TempDir(tmpBaseDir, "docker2aci-")
+	if err != nil {
+		return nil, nil, err
+	}
+	defer os.RemoveAll(tmpParentDir)
+
+	copier := progressutil.NewCopyProgressPrinter()
+
+	resultChan := make(chan layer, len(layerIDs))
+	for i, layerID := range layerIDs {
+		// https://github.com/golang/go/wiki/CommonMistakes
+		i := i // golang--
+		layerID := layerID
+		go func() {
+			tmpDir, err := ioutil.TempDir(tmpParentDir, "")
+			if err != nil {
+				resultChan <- layer{
+					index: i,
+					err:   fmt.Errorf("error creating dir: %v", err),
+				}
+				return
+			}
+
+			layerFile, closer, err := rb.getLayerV2(layerID, dockerURL, tmpDir, copier)
+			if err != nil {
+				resultChan <- layer{
+					index: i,
+					err:   fmt.Errorf("error getting the remote layer: %v", err),
+				}
+				return
+			}
+			resultChan <- layer{
+				index:  i,
+				file:   layerFile,
+				closer: closer,
+				err:    nil,
+			}
+		}()
+	}
+	for i := 0; i < len(layerIDs); i++ {
+		res := <-resultChan
+		defer res.closer.Close()
+		if res.err != nil {
```

Comment:
If an error occurs at `i`, we might end up with response bodies not closed for all layers with `i < index < len(layerIDs)`